### PR TITLE
Data logger tool

### PR DESCRIPTION
**Pull request by IonutMuthi**
Adding data logger tool

### CODE REVIEW

File: src/customScale.cpp [hunk 1]
```diff
@@ -60,8 +63,28 @@ void CustomScale::stop()
 	scaler->stopTimer();
 }
 
+int numDigits(double num){
+    if(int(num) == 0) {return -1; }
```
Comment (adisuciu) [inline src/customScale.cpp:67]: spacings

**Review by adisuciu — CHANGES_REQUESTED** [review]

File: src/gui/customcolqgridlayout.cpp [hunk 2]
```diff
@@ -0,0 +1,241 @@
...
+	}
+}
+
+void CustomColQGridLayout::computeCols(double width)
+{
+	int colCount = currentNumberOfCols;
+	auto availableWidth = this->width();
+
+	if (width != 0) {
+		colCount = availableWidth / width - 1;
```
Comment (AlexandraTrifan) [inline src/gui/customcolqgridlayout.cpp:187]: what happens when width == 1? we'll get some undefined behaviour..

File: src/gui/customcolqgridlayout.cpp [hunk 3]
```diff
@@ -0,0 +1,241 @@
...
+		colWidth = width;
+	}else{
+		if(m_widgetList.size() > 0){
+			colCount = availableWidth / m_widgetList.at(0)->width() - 1;
```
Comment (AlexandraTrifan) [inline src/gui/customcolqgridlayout.cpp:191]: same as above, I guess

File: src/gui/datalogger.cpp [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+void DataLogger::dataLoggerToggled(bool toggled)
+{
+	dataLoggerUI->toggleDataLoggerSwitch(toggled);
+	if (isRunningOn && !toggled) {
+		dataLoggerUI->disableDataLogging(true);
+		Q_EMIT isDataLogging(false);
+	}
+}
+
+void DataLogger::atteptDataLogging(){
```
Comment (AlexandraTrifan) [inline src/gui/datalogger.cpp:91]: typo in function name

File: src/datalogger/datalogger_api.cpp [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+
+int DataLogger_API::get_precision()
+{
+	return dataLogger->getPrecision();
+}
+
+void DataLogger_API::set_precision(int precision)
+{
+	dataLogger->setPrecision(precision);
+	Q_EMIT dataLogger->PrecisionChanged(precision);
```
Comment (AlexandraTrifan) [inline src/datalogger/datalogger_api.cpp:25]: can this be lowerCamelCase like all the others?

File: src/datalogger/datalogger_api.cpp [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+}
+
+int DataLogger_API::get_recording_interval()
+{
+	return dataLogger->getValueReadingTimeInterval();
+}
+
+void DataLogger_API::set_recording_interval(int interval)
+{
+	Q_EMIT dataLogger->RecordingIntervalChanged(interval);
```
Comment (AlexandraTrifan) [inline src/datalogger/datalogger_api.cpp:35]: same as above

File: src/datalogger/dataloggergenericmenu.cpp [hunk 7]
```diff
@@ -0,0 +1,165 @@
...
+	Q_EMIT changeHistorySize(numSamplesFromIdx(idx));
+}
+
+double DataLoggerGenericMenu::numSamplesFromIdx(int idx)
+{
+	switch(idx) {
+	case 0:	return 1.0;
+	case 1:	return 10.0;
+	case 2:	return 60.0;
+	default:throw std::runtime_error("Invalid INDEX");
```
Comment (AlexandraTrifan) [inline src/datalogger/dataloggergenericmenu.cpp:150]: are we handling the thrown error or Scopy will just crash?

File: src/dataloggermodel.h [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+
+class DataLoggerModel: public QObject
+{
+	Q_OBJECT
+
+public:
+	DataLoggerModel();
+	~DataLoggerModel();
+	void createChannel(QString name, CHANNEL_DATA_TYPE type, CHANNEL_FILTER filter);
+	void distroyChannel(QString name);
```
Comment (AlexandraTrifan) [inline src/dataloggermodel.h:40]: typo in function name?

File: src/gui/channel_manager.cpp [hunk 9]
```diff
@@ -29,66 +63,61 @@ ChannelManager::ChannelManager(ChannelsPositionEnum position, QWidget* parent)
...
 	ChannelWidget* ch = new ChannelWidget(chId, deletable, simplefied, color);
-
-	m_channelsWidget->layout()->addWidget(ch);
-	ch->setFullName(fullName + QString(" %1").arg(chId + 1));
-	ch->setShortName(shortName + QString(" %1").arg(chId + 1));
+	m_channelsWidget->layout()->setMargin(0);
+    m_channelsWidget->layout()->addWidget(ch);
+	if (m_channelIdVisible) {
+        ch->setFullName(fullName + QString(" %1").arg(chId + 1));
+        ch->setShortName(shortName + QString(" %1").arg(chId + 1));
```
Comment (AlexandraTrifan) [inline src/gui/channel_manager.cpp:96]: formatting is still not good

File: src/gui/dataloggercontroller.cpp [hunk 10]
```diff
@@ -0,0 +1,139 @@
...
+	if (uiFilter == "Average") {
+		filter = CHANNEL_FILTER::AVERAGE;
+	}
+	if (uiFilter == "All") {
+		filter = CHANNEL_FILTER::ALL;
+	}
+	dataLoggerModel->createChannel(name,type,filter);
+}
+
+void DataLoggerController::distroyChannel(QString name)
```
Comment (AlexandraTrifan) [inline src/gui/dataloggercontroller.cpp:41]: typo in function name?

File: src/gui/dataloggerview.cpp [hunk 11]
```diff
@@ -0,0 +1,217 @@
...
+			dataLoggingFilePath->setText(tr("No file selected"));
+			dataLoggingFilePath->setStyleSheet("color:red");
+		} else {
+			dataLoggingFilePath->setStyleSheet("color:white");
+			Q_EMIT pathChanged(path);
+		}
+	});
+
+	dataLoggingLayout->addWidget(dataLoggingFilePath);
+	dataLogginBrowseBtn = new QPushButton("Browse");
```
Comment (AlexandraTrifan) [inline src/gui/dataloggerview.cpp:51]: missing character in name?

File: src/lcdNumber.hpp [hunk 12]
```diff
@@ -35,6 +35,8 @@ namespace adiscope {
 
 		public Q_SLOTS:
 			void display(double num);
+            void setPrecision(unsigned precision);
+            unsigned getPrecision();
```
Comment (AlexandraTrifan) [inline src/lcdNumber.hpp:39]: bad indent

File: src/sismograph.hpp [hunk 13]
```diff
@@ -51,25 +53,47 @@ namespace adiscope {
...
 		QVector<double> ydata;
-		CustomFifo<double> xdata;
+		QVector<double> xdata;
+
+		void updateScale();
+		double findMaxInFifo();
+		bool autoscale;
+
+	Q_SIGNALS:
+        void dataChanged(std::vector<double> data);
```
Comment (AlexandraTrifan) [inline src/sismograph.hpp:96]: bad indent

File: src/tool.cpp [hunk 14]
```diff
@@ -72,7 +73,9 @@ Tool::~Tool()
 {
 	disconnect(prefPanel, &Preferences::notify, this, &Tool::readPreferences);
 
-	run_button->setChecked(false);
+	if (run_button) {
+        run_button->setChecked(false);
```
Comment (AlexandraTrifan) [inline src/tool.cpp:77]: bad indent

**Review by AlexandraTrifan — CHANGES_REQUESTED** [review]

**Pull request merged**